Codex CLI Support: Emit strict object-shaped input schemas for MCP tools

alexmahrou opened the pull request:
**Summary**
- Generated dynamic Pydantic models for every tool, emitting strict Draft‑07 object schemas with additionalProperties: false and eliminating "type": "integer" across the board.
- Added a lightweight tools/list shim to coerce any legacy descriptors into compliant schemas.

There are around 10's of millions of ChatGPT Plus users who now have the ability to sign in with their ChatGPT Credentials in the Codex CLI at no additional cost or API token hit. Providing compatibility with this platform will provide another onramp into QC Ecosystem. The changes tighten every tool’s input schema so it is a Draft‑07 JSON object with additionalProperties: false and no "integer" types. By returning schemas that conform to the same spec most MCP clients expect, OpenAI’s Codex, Anthropic’s Claude, and any other MCP-compliant client can parse arguments deterministically without needing client‑specific shims or guessing field types. This reduces “missing field 'type'” errors and ensures consistent validation for all clients, not just Codex.


I've tested this branch locally with Codex CLI and it works. 

Related to [https://github.com/QuantConnect/mcp-server/issues/21](https://github.com/QuantConnect/mcp-server/issues/21)

jaredbroad commented:
@alexmahrou We tested this with codex and it still had issues. Perhaps there's some nuance about your setup that made it work for you? Can you please share a screen recording? We'd love to make it work with Codex.

alexmahrou committed:
Emit object-shaped MCP tool schemas for Codex/Claude compatibility

alexmahrou commented:
Hi @jaredbroad ,

Thanks for taking another look at this. I rebased my branch from the latest upstream `main` this morning and confirmed everything still works end-to-end with Codex CLI. My branch is only one commit ahead, so it should be convenient to merge once we iron this out.

To mirror the video you shared, I recorded a YouTube walkthrough where Codex selects and runs a Dual Momentum Rotation algorithm:

▶️ [[YouTube: Codex + QC MCP Server Hello World](https://youtu.be/iBQ8uicyKJ0)](https://youtu.be/iBQ8uicyKJ0)

And here’s the backtest Codex produced during the run:

🔗 [[View QuantConnect Backtest](https://www.quantconnect.cloud/backtest/d9a43212636591b2dde566388b427ef9/?theme=darkly)](https://www.quantconnect.cloud/backtest/d9a43212636591b2dde566388b427ef9/?theme=darkly)

---

## 🔧 Repro Instructions (Hello World with Codex CLI)

### 1. Clone & checkout my branch

```bash
mkdir -p ~/work && cd ~/work
git clone git@github.com:alexmahrou/mcp-server.git
cd mcp-server
git checkout codex/fix-mcp-tool-schemas-for-codex-compatibility
```

### 2. Python environment

```bash
python3 -m venv .venv
. .venv/bin/activate
pip install -U pip
pip install -e ".[test]"
```

### 3. Codex installation

Codex CLI must be installed (`npm i -g @codex/cli`), then you can run it directly with:

```bash
DEBUG=* codex
```

### 4. Codex config

Codex looks for `~/.codex/config.toml`. Here’s the working format:

```toml
[projects."/home/alexm"]
trust_level = "trusted"

[projects."/home/alexm/work/mcp-server"]
trust_level = "trusted"

[mcp_servers.quantconnect]
command = "bash"
args = [
  "-lc",
  "cd /home/alexm/work/mcp-server && exec /home/alexm/work/mcp-server/.venv/bin/python -m src.main"
]

[mcp_servers.quantconnect.env]
QUANTCONNECT_USER_ID = "<your_user_id>"
QUANTCONNECT_API_TOKEN = "<your_api_token>"
```

Replace the env values with your QC credentials.

### 5. Connect from Codex

Inside Codex chat:

```
/status
connect to the "quantconnect" MCP server
```

Expected working tool calls:

```
quantconnect.read_mcp_server_version {}
quantconnect.read_account {}
quantconnect.list_projects {}
```

---

This is the minimal repro I tested with. Codex CLI successfully:

* Created a new QC project
* Added a minimal algorithm file
* Compiled
* Launched a backtest (results linked above)

I tried to keep this setup guide in parity with your published video, so anyone should be able to follow it 1:1 and reproduce Codex driving the QC MCP server.

Let me know if you’d like me to add this quickstart section to the README on my branch so future contributors can get Codex working out-of-the-box.

DerekMelchin commented:
Hi @alexmahrou 

The video above shows the agent commonly struggles to provide the correct arguments to each tool and some tools still show "tool failed". 

We've created a GitHub Issue on the Codex repo to get the underlying problem addressed/pointed out.
https://github.com/openai/codex/issues/2420

alexmahrou commented:
Oh sweet good to see this. Curious did you try launching codex with 
```
DEBUG=* codex
```
?

Maybe that would give us more verbose logging?

DerekMelchin commented:
I used the most verbose logging I could
```
$env:RUST_LOG = "codex_core=trace,codex_tui=trace"
```
but the logs still didn't elaborate on the problem.

DerekMelchin commented:
Likely related to https://github.com/openai/codex/issues/3152
